File: examples/higher-order-functions/map.rs
fn map<T, U>(
  arr: Vec<T>,
  f: &Fn(&T) -> U,
) -> Vec<U> {
  let mut result = Vec::new();
  for item in arr.iter() {
    result.push(f(item))
  }
  return result;
}

fn main() {
  let new_vec = map(vec![1, 2, 3], &|n| n * 2);
  print!("{:?}", new_vec);
}

File: examples/higher-order-functions/map.kt
fun <T, U>map(
  arr: List<T>,
  f: (item: T) -> U
): List<U> {
  var result = mutableListOf<U>()
  for (item: T in arr) {
    result.add(f(item))
  }
  return result
}

fun main() {
  var newList = map(
    listOf(1, 2, 3),
    { n : Int -> n * 2 }
  )
  print(newList)
}
